onnxruntime/core/providers/qnn/qnn_execution_provider.cc

```diff
@@ -706,4 +739,147 @@ const InlinedVector<const Node*> QNNExecutionProvider::GetEpContextNodes() const
 
   return ep_context_nodes;
 }
+
+QNNExecutionProvider::PerThreadContext::PerThreadContext(qnn::QnnBackendManager* qnn_backend_manager,
+                                                         uint32_t device_id,
+                                                         uint32_t core_id,
+                                                         qnn::HtpPerformanceMode default_htp_performance_mode,
+                                                         uint32_t default_rpc_control_latency)
+    : qnn_backend_manager_(qnn_backend_manager) {
+  Status rt = qnn_backend_manager_->CreateHtpPowerCfgId(device_id, core_id, htp_power_config_id_);
+  is_htp_power_config_id_valid_ = rt.IsOK();
+  // default_htp_performance_mode and default_rpc_control_latency are from QNN EP option.
+  // set it once only for each thread as default so user don't need to set it for every session run
+  if (is_htp_power_config_id_valid_) {
+    if (qnn::HtpPerformanceMode::kHtpDefault != default_htp_performance_mode) {
+      ORT_IGNORE_RETURN_VALUE(qnn_backend_manager_->SetHtpPowerConfig(htp_power_config_id_,
```

Comment:
I assume all these api's SetHtpPower, SetRpcControl, DestoryHtpPowerConfigId etc. which eventually call qnn api's are all thread safe?

Reply:
good point, haven't run into any issue with that. But need to double confirm.

---
In reply to: [1493055792](https://github.com/microsoft/onnxruntime/pull/19521#discussion_r1493055792) [](ancestors = 1493055792)

Reply:
Confirmed. Yes, they are thread safe.